### DISCUSSION TITLE
Run activities as soon or as late as possible in a scheduling problem

### DISCUSSION THREAD

**Discussion by nicola-gigante** [aiplan4eu/unified-planning#641] [Q&A]
Hi!

I'm solving `SchedulingProblem`s with the Aries backend and everything works smoothly.

However, now I have the need of optimizing the plan so that activities are executed globally as soon as possible.
This is different from optimizing the total makespan, because I have some long running activities that contribute mostly to the makespan while other smaller, concurrent activities are placed arbitrarily in the middle. Those are the ones that I would like to pull as early as possible.

In a different scenario I need to do the same thing but pushing the actions as late as possible within their deadlines.

At first I tried to use `MinimizeExpressionOnFinalState` and `MaximizeExpressionOnFinalState`, to minimize or maximize the sum of the starting/ending points of the activities, but it does not work because the expression has to be built on top of numeric fluents, and timepoints are not accepted.

So is there currently a way to achieve this?

**Comment by arbimo** [comment]
> Hi!
> 
> I'm solving `SchedulingProblem`s with the Aries backend and everything works smoothly.

Glad to know!

Regarding the issue at end, it is indeed a limitation currently. I think it would be important to address to represent common objective function (notably earliness and tardiness based metrics).
When adding the support of time in expression, we took the decision of time expressions (i.e. value of timepoints) to have a type distinct from int or reals.

One very pragmatic reason is that its underlying representation (int or real) will differ depending on whether the problem considers continuous or discrete time.

So we have three types of numeric values: int, real and time. All previous numeric expression handling were focused on int or real, raising an exception if one of the arguments were not in this realm. I think in most cases, they could be relaxed, i.e., allowing a time expression were we previously allowed only int or real. The place where care should be taken however is on the definition of the resulting type.


To illustrate that, here is a list of a few expression and their possible types:

- 3 : Int
- 3.5 : Real
- end(t1) : Time
- end(t1) - start(t1) : Time 
- 3 * end(t1) : Time 
- 3.5 * end(t1) : Time ???? (this is still a time but you may loose the capability to represent it with ints)
- end(t1) * end(t1) : ??? (Time²)

**Reply by nicola-gigante** [reply]
Hi @arbimo!

I see, the issue is not simple!

I would say that multiplying two times should indeed be "time squared".
Some simple sort of dimensionality analysis should not be too hard to do, to forbid nonsensical comparisons (like time <= time squared). 

But otherwise, I think it may make sense to multiply two times if you are trying to model some physical quantity you want to optimize. Acceleration is space/time squared, for example. 

This actually opens a Pandora's box but would it be cool if we could give physical units to the quantities represented by fluents?
Like time points would explicitly be represented as Python's `datetime` or `timedelta` and energy consumed by a robot's battery could explicitly be measured in kWh or Joules. Just thinking aloud...
